### PR TITLE
Switch controller-manager to logr (1)

### PR DESCRIPTION
**Pull request by timebertt**
**How to categorize this PR?**

/area logging dev-productivity
/kind enhancement

**What this PR does / why we need it**:

This is the first PR of a few, that switches the controllers in gardener-controller-manager from `logrus` to `logr`.
It tackles:
- harmonizing the log format (json) of logrus and zap to make logs less confusing during migration period
- adding zap logr in gcm and gardenlet
- migrating logrus usages in common packages (shared between gcm and gardenlet)
- migrating the first few gcm controllers

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener/issues/4251

**Special notes for your reviewer**:
/squash

I noticed, that there will be some common questions about different logging aspects. So I plan to add a developer guide answering these questions in one of the following PRs, basically a condensed version of
- https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md#structured-logging-in-kubernetes
- https://github.com/go-logr/logr
- https://dave.cheney.net/2015/11/05/lets-talk-about-logging

**Release note**:

```other operator
gardener-controller-manager and gardenlet have started switching from logrus to zap. Make sure to use the `json` log format to have harmonized logging during the migration period.
```

### CODE REVIEW

File: cmd/gardener-controller-manager/app/gardener_controller_manager.go [hunk 1]
```diff
@@ -204,18 +206,33 @@ func NewGardener(ctx context.Context, cfg *config.ControllerManagerConfiguration
...
 		if err := flag.Value.Set(fmt.Sprintf("%d", cfg.KubernetesLogLevel)); err != nil {
 			return nil, err
 		}
 	}
 
+	if _, err := maxprocs.Set(maxprocs.Logger(func(s string, i ...interface{}) {
+		log.Info(fmt.Sprintf(s, i...))
+	})); err != nil {
+		log.Error(err, "failed to set GOMAXPROCS")
+	}
```
Comment (stoyanr) [inline cmd/gardener-controller-manager/app/gardener_controller_manager.go:237]: With this change, we are replacing `automaxprocs` with `maxprocs`. Is this just done for purposes of logging, or is there a change in the actual `GOMAXPROCS` value that is being set? 
Also, could you please add a small comment above this section explaining what it's doing as it's at least for me pretty hard to understand what is going on - it looks like we are setting a logger, but in fact from the log message it seems we are actually setting `GOMAXPROCS`?

File: pkg/controllerutils/worker.go [hunk 2]
```diff
@@ -20,38 +20,48 @@ import (
...
-func CreateWorker(ctx context.Context, queue workqueue.RateLimitingInterface, resourceType string, reconciler reconcile.Reconciler, waitGroup *sync.WaitGroup, workerCh chan<- int, injectFn ...inject.Func) {
-	fns := append(injectFn, func(i interface{}) error {
-		_, err := inject.StopChannelInto(ctx.Done(), i)
-		return err
-	})
-
-	for _, f := range fns {
-		if err := f(reconciler); err != nil {
-			logger.Logger.Errorf("An error occurred while reconciler injection: %v", err)
-		}
```
Comment (stoyanr) [inline pkg/controllerutils/worker.go:47]: Why has this code been removed? Is there a controller-runtime change that makes it obsolete (or just not a good idea)?
Do I understand correctly that this change is not really related to logging? If yes, is it a good idea to have it in this PR?

File: pkg/apis/core/v1alpha1/helper/helper.go [hunk 3]
```diff
@@ -659,18 +658,6 @@ func GetShootMachineImageFromLatestMachineImageVersion(image gardencorev1alpha1.
...
-	for _, machineImage := range machineImages {
-		for idx, worker := range workers {
-			if worker.Machine.Image != nil && machineImage.Name == worker.Machine.Image.Name {
-				logger.Logger.Infof("Updating worker images of worker '%s' from version %s to version %s", worker.Name, *worker.Machine.Image.Version, *machineImage.Version)
-				workers[idx].Machine.Image = machineImage
-			}
-		}
-	}
-}
-
```
Comment (stoyanr) [inline pkg/apis/core/v1alpha1/helper/helper.go:673]: Why was this function removed - I assume it showed up as dead code? But then, the same seems to be true for the same function in `pkg/apis/core/v1beta1/helper/helper.go`?

File: pkg/scheduler/controller/shoot/reconciler.go [hunk 4]
```diff
@@ -26,46 +26,45 @@ import (
...
 type reconciler struct {
-	logger       logr.Logger
 	config       *config.ShootSchedulerConfiguration
 	gardenClient client.Client
 	recorder     record.EventRecorder
 }
 
 func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log := logf.FromContext(ctx)
+
```
Comment (stoyanr) [inline pkg/scheduler/controller/shoot/reconciler.go:51]: How are we sure the log retrieved from the context is not nil, that is, where is the matching `IntoContext`? I only see `IntoContext` being used in `CreateWorker`, but it seems the scheduler is not using that function? I assume we are relying on some `controller-runtime` magic that always injects a logger in worker contexts?

**Review by stoyanr — CHANGES_REQUESTED** [review]
Looks quite good from my perspective, my only concerns are regarding missing comments and changes that potentially don't belong to this PR. I also did some local testing and didn't encounter anything unusual (but I didn't test all components / controllers affected by this change).

File: pkg/controllerutils/worker.go (hunk 2 above)
Comment (timebertt) [reply pkg/controllerutils/worker.go:47]: I removed it because this code was unused and it was easier to do my changes (optionally injecting a logger) by removing this first.

File: pkg/apis/core/v1alpha1/helper/helper.go (hunk 3 above)
Comment (timebertt) [reply pkg/apis/core/v1alpha1/helper/helper.go:673]: Yep, this was unused (usage was removed in #4438). I simply removed it here because it used the global `logger.Logger`.
I deleted the one in `pkg/apis/core/v1beta1/helper/helper.go` now as well.

File: pkg/scheduler/controller/shoot/reconciler.go (hunk 4 above)
Comment (timebertt) [reply pkg/scheduler/controller/shoot/reconciler.go:51]: The counterpart is in the controller-runtime controller implementation: https://github.com/gardener/gardener/blob/8abb827821db42806a828c66b93995e1a54c693c/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go#L112-L113

The scheduler already uses this code and the other components will use this as well once they migrated to the manager. `CreateWorker` now does the same to make it easy to switch later on without thinking about logging.

BTW, `logf.FromContext` will never return `nil`. If the context is `nil` or doesn't carry a logger, the func will simply return the global `logf.Log` (which is also never `nil`, but might be a `NullLogger` if not initialized (e.g. in tests)).

File: cmd/gardener-controller-manager/app/gardener_controller_manager.go (hunk 1 above)
Comment (timebertt) [reply cmd/gardener-controller-manager/app/gardener_controller_manager.go:237]: Valid point, this is not very obvious. I added a comment explaining what's happing there.

**Comment by timebertt** [comment]
@stoyanr Thank you for your review.
I addressed your feedback in a new commit, PTAL :)

**Review by stoyanr — APPROVED** [review]
Thanks for the explanations!

/lgtm

**Review by rfranzke — APPROVED** [review]
/lgtm
